Phase-1-Practices/Week-4/Julian-P-Nguyen/README.md

```diff
@@ -0,0 +1,729 @@
+metadata:
+  name: wordpress
+  labels:
+    app: wordpress-k8s
+spec:
+  selector:
+    matchLabels:
+      app: wordpress-k8s
+      tier: frontend
+  strategy:
+    type: Recreate
+  template:
+    metadata:
+      labels:
+        app: wordpress-k8s
+        tier: frontend
+    spec:
+      containers:
+      - name: wordpress
+        image: bitnami/wordpress:latest
+        env:
+          - name: ALLOW_EMPTY_PASSWORD
+            value: yes
+          - name: WORDPRESS_DATABASE_USER
+            valueFrom:
+              secretKeyRef:
+                name: secret-wordpress
+                key: db_user
+          - name: WORDPRESS_DATABASE_NAME
+            valueFrom:
+              secretKeyRef:
+                name: secret-wordpress
+                key: db_name
+          - name: WORDPRESS_DATABASE_PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: secret-wordpress
+                key: db_password
+        ports:
+        - containerPort: 80
+          name: wordpress
+        volumeMounts:
+        - name: wordpress-storage
+          mountPath: /var/www/html
+      volumes:
+      - name: wordpress-storage
+        persistentVolumeClaim:
+          claimName: wordpress-volume
+````
+
+### **DEPLOY APPLICATIONS WITH `Kubernetes`**:
+
+- Create `mariadb` Secret object:
+
+````bash
+$ kubectl apply -f ./secret-mariadb.yml
+````
+
+<img src="./imgs/create-mariadb-secret.png">
+
+- Create `wordpress` Secret object:
+
+````bash
+$ kubectl apply -f ./secret-wordpress.yml
+````
+
+<img src="./imgs/create-wordpress-secret.png">
+
+- Verify added secrets existence:
+````bash
+
+$ kubectl get secrets
+````
+
+- Deploy `mariadb`:
+````bash
+
+$ kubectl apply -f ./mariadb-deployment.yml
+````
+
+<img src="./imgs/service-persistent-deployment-mariadb.png">
+
+
+- Deploy `wordpress`:
+````bash
+
+$ kubectl apply -f ./wordpress-deployment.yml
+````
+
+<img src="./imgs/service-persistent-deploynent-wordpress.png">
+
+
+- Check `Pod`(s):
+
+````bash
+$  kubectl get pods
+````
+
+<img src="./imgs/running-pods.png">
+
+- Check `Service`:
+
+````bash
+$  kubectl get services (<Service-name>)
+````
+
+**Note**
+> `EXTERNAL-IP` always `<Pending>` because `Minikube` exposes Services through `NodePort` only.
+
+
+<img src="./imgs/services-list.png">
+
+
+## **C. ACCESS `Wordpress` page:**
+- Check IP of `wordpress`:
+
+```bash
+$ minikube service wordpress --url
+```
+
+<img src="./imgs/view-url-wordpress.png">
+
+
+- Visit Landing page:
+
+```bash
+$ curl http://<CLUSTER-IP>:<High-PORT>
+```
+
+:heavy_check_mark: **Exptected outcome:**
+
+<img src="./imgs/success-landing-page-curl.png">
+
+# **V. TROUBLESHOOTING:**
+
+## **`MUST-KNOW` DEBUG COMMANDS**
+
+- Entering a `Container` in K8S with `bash`:
+
+```bash
+$ kubectl exec --stdin --tty <Pod-name> -- /bin/bash
+```
+
+- View logs:
+
+```bash
+$ kubectl logs <pod-name>
+```
+
+## 1. `Deployment in version "v1" cannot be handled as a Deployment...`
+
+:x: Bug detected:
```

Comment:
Đây không phải bug, mà chỉ là 1 gotcha về cách hoạt động của yaml mà thôi.

Nếu em còn nhớ phần thực hành ansible thì trong docs có ghi chú đại khái là: với các giá trị boolean nhưng ở dạng text (yes/no, true/false, ...) mà muốn giữ nguyên dạng text thì phải đặt trong ngoặc, nếu không nó sẽ bị auto convert thành kiểu boolean luôn.

Reply:
Dạ! Em sẽ edit lại đoạn này thêm vào commit lại. Thank you for your feedback ạ!

Reply:
Em đã hoàn thiện edit lại đoạn này ạ. Anh @tovin07 review qua giúp em nhé! Em cám ơn anh.